text/0000-edition-2021-and-beyond.md

```diff
@@ -0,0 +1,196 @@
+
+## Key ideas of edition do not change
+
+Just as with Rust 2018, we are firmly committed to the core concepts of an edition:
+
+* Crates using older editions continues to compile in the newer
+  compiler, potentially with warnings.
+* Crates using different editions can interoperate, and people can
+  upgrade to newer editions on their own schedule.
+* Code that compiles without a warning on Edition N should also
+  compile on Edition N + 1.
+* Migration between editions should generally be automated.
+* Editions make "skin-deep" changes, with all editions ultimately
+  compiling to a single common representation.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+We use this section to try and convey the story that average users will need to understand.
+
+## What is a Rust edition?
+
+Every three years, we introduce a new Rust Edition. These editions are named after the year in which they occur, like Rust 2015 or Rust 2018. Each crate specifies the Rust edition that it requires in its `Cargo.toml` file via a setting like `edition = "2018"`. The purpose of editions is to give us a chance to introduce "opt-in" changes like new keywords that would otherwise have the potential to break existing code.
+
+When we introduce a new edition, we don't remove support for the older ones, so all crates continue to compile just as they ever did. Moreover, editions are fully interoperable, so there is no possibility of an "ecosystem split". This means that you can upgrade your crates to the new edition on whatever schedule works best for you.
+
+## How do I upgrade between editions?
+
+Upgrading between editions is meant to be easy. The general rule is, if your code compiles without warnings, you should be able to opt into the new edition, and your code will compile.
+
+Along with each edition, we also release support for it in a tool called `rustfix`, which will automatically migrate your code from the old edition to the new edition, preserving semantics along the way. You may have to do a bit of cleanup after the tool runs, but it shouldn't be much.
+
+## "Migrations" in an edition vs "idiom lints"
+
+When we release a new edition, it comes together with a certain set of "migrations". Migrations are the "breaking changes" introduced by the edition, except of course that since editions are opt-in, no code actually breaks. For example, if we introduce a new keyword, you will have to rename variables or functions using the old keyword, or else use Rust's `r#keyword` feature (which allows you to use a keyword as a regular variable/function/type name). As mentioned before, the edition comes with tooling that will make these changes for you, though sometimes you will want to cleanup the resulting code afterwards.
```

Comment:
A thing worth highlighting somehow:

 - Migration lints take code that compiles on edition X and turn it into code that compiles on edition X AND X+1
 - Idiom lints take code that compiles on edition X+1 and turn it into code that is idiomatic for edition X+1 (but may no longer compile on edition X)

Reply:
how would "but may no longer compile on edition X" work? Code shouldn't suddenly "no longer compile on &lt;specification from the past&gt;"

Reply:
I had to read that twice too. I think what @Manishearth meant is that -- once you've applied the idiom lint, the resulting code (in the new edition) may not compile in the old edition anymore. 

To be honest, I had forgotten that migrations are supposed to target the "intersection" of editions, so this is perhaps a good addition -- but I'm not sure if it's *always* true. I remember thinking that we might have to have cases were the migration required you to use the new edition. But I guess we avoided those for the module transition? I'd be ok writing the above as the general rules -- kind of a "unless otherwise stated" sort of thing.

Reply:
@nikomatsakis The migration/idiom split is _deliberately_ designed to deal with this specific problem. Idiom lints do not necessarily produce code that does not compile on older editions, but typically do. That's more central to their working than "will become on by default in the next edition" actually.

> how would "but may no longer compile on edition X" work? Code shouldn't suddenly "no longer compile on <specification from the past>"

I'm talking about the migration present within those lints.

Reply:
The reasoning here is that it is not always possible to migrate wholesale, so the edition lints migrate it to be _compatible_ with the new edition (while still compiling on the old one so that you can perform any manual fixes). The idiom lints finish off any parts of the migration and apply potentially incompatible changes that make the code more idiomatic (but are not necessary to make the code compile).

Reply:
Yeah, this is sounding quite familiar. I will add some text to that effect.

Reply:
However, I believe that idiom lints can also be used to change the default severity from warning (older editions) to error (newer editions) correct? I'll have to review the code, but I'm pretty sure that's something we also want.

Reply:
OK, I pushed some changes -- @Manishearth please take a look and see if this sounds right to you.